File: server/bcephal-reco/src/main/java/com/moriset/bcephal/reconciliation/domain/WriteOffFieldValue.java
/**
 * 
 */
package com.moriset.bcephal.reconciliation.domain;

import java.math.BigDecimal;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.moriset.bcephal.domain.Persistent;
import com.moriset.bcephal.domain.filters.PeriodValue;

import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PostLoad;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/**
 * @author Joseph Wambo
 *
 */
@Entity(name = "WriteOffFieldValue")
@Table(name = "BCP_WRITE_OFF_FIELD_VALUE")
@Data
@EqualsAndHashCode(callSuper = false)
public class WriteOffFieldValue extends Persistent {

	private static final long serialVersionUID = 3237993022227596122L;

	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "writeoff_field_value_seq")
	@SequenceGenerator(name = "writeoff_field_value_seq", sequenceName = "writeoff_field_value_seq", initialValue = 1,  allocationSize = 1)
	private Long id;
	
	@ToString.Exclude
	@EqualsAndHashCode.Exclude	
	@JsonIgnore
	@ManyToOne(fetch = FetchType.LAZY)
	@jakarta.persistence.JoinColumn(name = "field")
	private WriteOffField field;
	
	private int position;
	
	private boolean defaultValue;
		
	private String stringValue;	
	private BigDecimal decimalValue;	
	@Embedded	
	private PeriodValue dateValue;
	
	
	@JsonIgnore
	@Override
	public WriteOffFieldValue copy() {
		WriteOffFieldValue copy = new WriteOffFieldValue();
		copy.setPosition(position);
		copy.setDefaultValue(defaultValue);
		copy.setStringValue(stringValue);
		copy.setDecimalValue(decimalValue);
		copy.setDateValue(dateValue.copy());
		return copy;
	}
	
	public WriteOffFieldValue() {
		dateValue = new PeriodValue();
	}
	
	
	public void setDateValue(PeriodValue dateValue) {
		if(dateValue == null) {
			this.dateValue = new PeriodValue();
		}else {
			this.dateValue = dateValue;
		}
	}
	
	@PostLoad
	public void initListChangeHandler() {
		if(dateValue == null) {
			 dateValue = new PeriodValue();
		}
	}

}

File: client/bcephal-client-model/Reconciliation/WriteOffFieldValue.cs
using Bcephal.Models.Base;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Bcephal.Models.Reconciliation
{
    public class WriteOffFieldValue : Persistent
    {

		public int Position { get; set; }

		public bool DefaultValue { get; set; }

		public string StringValue { get; set; }

		public decimal? DecimalValue { get; set; }
		
		public PeriodValue DateValue { get; set; }

		public WriteOffFieldValue()
		{
			
		}

		public override int CompareTo(object obj)
		{
			if (obj == null || !(obj is WriteOffFieldValue)) return 1;
			if (this == obj) return 0;
			if (this.Id.HasValue && this.Id.Equals(((WriteOffFieldValue)obj).Id)) return 0;
			if (this.Position.Equals(((WriteOffFieldValue)obj).Position) && !string.IsNullOrWhiteSpace(this.StringValue))
			{
				return this.StringValue.CompareTo(((WriteOffFieldValue)obj).StringValue);
			}
			return this.Position.CompareTo(((WriteOffFieldValue)obj).Position);
		}

		[JsonIgnore]
		public string Key
		{
			get
			{
				if (string.IsNullOrWhiteSpace(Key_))
				{
					Key_ = Guid.NewGuid().ToString("d");
				}
				return Key_;
			}
			set
			{
				Key_ = value;
			}
		}
		[JsonIgnore]
		private string Key_ { get; set; }

	}
}
